File: julia/particle.jl
module MyParticle


export Particle, Force

type Particle{T<:Real}
    position::Force{T}
    velocity::Force{T}
    acceleration::Force{T}
    #jerk::Force{T}
    #energy::T
    mass::T
    time::T
end

type Force{T<:Real}
    x::T
    y::T
    z::T
end

Force{T<:Real}(x::T) = Force{T}(x, x, x)

norm2{T<:Real}(f::Force{T}) = f.x*f.x + f.y*f.y + f.z*f.z
norm{T<:Real}(f::Force{T}) = sqrt(f.x*f.x + f.y*f.y + f.z*f.z)

+{T<:Real}(a::Force{T}, b::Force{T}) = Force{T}(a.x+b.x, a.y+b.y, a.z+b.z)
-{T<:Real}(a::Force{T}, b::Force{T}) = Force{T}(a.x-b.x, a.y-b.y, a.z-b.z)
*{T<:Real}(a::Force{T}, b::Force{T}) = a.x*b.x, a.y*b.y, a.z*b.z

*{T<:Real}(a::Force{T}, b::T) = Force{T}(a.x*b, a.y*b, a.z*b)
*{T<:Real}(b::T, a::Force{T}) = b*a
/{T<:Real}(a::Force{T}, b::T) = Force{T}(a.x/b, a.y/b, a.z/b)

end # module

File: java/nbody/Particle.java
package nbody;

public class Particle 
{
    Force position;
    Force velocity;
    Force acceleration;
    double mass;

    Particle(){
        position = new Force(0., 0., 0.);
        velocity = new Force(0., 0., 0.);
        acceleration = new Force(0., 0., 0.);
        mass = 0.;
    }

    Particle(Force r, Force v, double m) {
        position = r;
        velocity = v;
        acceleration = new Force(0., 0., 0.);
        mass = m;
    }

    public void calcforce(Particle py) {
        Force sep = position.dist(py.position);

        double d2 = sep.norm2();
        d2 = -mass*py.mass/d2/Math.sqrt(d2);
        acceleration.x += d2 * sep.x;
        acceleration.y += d2 * sep.y;
        acceleration.z += d2 * sep.z;        
    }
}




